lib/fs.js

```diff
@@ -12,63 +12,68 @@ const DevMiddlewareError = require('./DevMiddlewareError');
-        const { writeToDisk: filter } = context.options;
-        let { outputPath } = compiler;
+      compiler.hooks.emit.tap('WebpackDevMiddleware', (compilation) => {
+        compiler.hooks.assetEmitted.tapAsync(
+          'WebpackDevMiddleware',
+          (file, content, callback) => {
+            let targetFile = file;
 
-        if (outputPath === '/') {
-          outputPath = compiler.context;
-        }
+            const queryStringIdx = targetFile.indexOf('?');
 
-        for (const assetPath of Object.keys(assets)) {
-          let targetFile = assetPath;
+            if (queryStringIdx >= 0) {
+              targetFile = targetFile.substr(0, queryStringIdx);
+            }
 
-          const queryStringIdx = targetFile.indexOf('?');
+            let { outputPath } = compiler;
 
-          if (queryStringIdx >= 0) {
-            targetFile = targetFile.substr(0, queryStringIdx);
-          }
+            // TODO Why? Need remove in future major release
+            if (outputPath === '/') {
+              outputPath = compiler.context;
+            }
 
-          const targetPath = path.isAbsolute(targetFile)
-            ? targetFile
-            : path.join(outputPath, targetFile);
-          const allowWrite =
-            filter && typeof filter === 'function' ? filter(targetPath) : true;
+            outputPath = compilation.getPath(outputPath, {});
 
-          if (allowWrite) {
-            const asset = assets[assetPath];
-            let content = asset.source();
+            const targetPath = path.join(outputPath, targetFile);
 
-            if (!Buffer.isBuffer(content)) {
-              // TODO need remove in next major release
-              if (Array.isArray(content)) {
-                content = content.join('\n');
-              }
+            const { writeToDisk: filter } = context.options;
+            const allowWrite =
+              filter && typeof filter === 'function'
+                ? filter(targetPath)
+                : true;
 
-              content = Buffer.from(content, 'utf8');
+            if (!allowWrite) {
+              return callback();
             }
 
-            mkdirp.sync(path.dirname(targetPath));
-
-            try {
-              fs.writeFileSync(targetPath, content, 'utf-8');
-
-              log.debug(
-                colors.cyan(
-                  `Asset written to disk: ${path.relative(
-                    process.cwd(),
-                    targetPath
-                  )}`
-                )
-              );
-            } catch (e) {
-              log.error(`Unable to write asset to disk:\n${e}`);
-            }
+            const { log } = context;
+            const dir = path.dirname(targetPath);
+
+            return mkdirp(dir, (mkdirpError) => {
+              if (mkdirpError) {
+                return callback(mkdirpError);
+              }
+
+              return fs.writeFile(targetPath, content, (writeFileError) => {
+                if (writeFileError) {
```

Comment:
Should we output `Unable to write asset to disk: xxx` as an error?

Reply:
I think it is unnecessary, we failed compilation if something not written as expected, because it is potentially break appication

Reply:
ok, agree with you.